### DISCUSSION TITLE
Fluent API and data return type in extensions

### DISCUSSION THREAD

**Discussion by franky47** [prisma/prisma#20901] [Q&A]
### Question

I'm maintaining an extension that performs [field-level encryption](https://github.com/47ng/prisma-field-encryption), by walking the tree of returned data to look for fields to decrypt.

Recently, I had a [bug report](https://github.com/47ng/prisma-field-encryption/issues/80) indicating that it doesn't work with the [Fluent API](https://www.prisma.io/docs/concepts/components/prisma-client/relation-queries#fluent-api).

The heart of the problem is that the query passed to the extension asks for a type of data, but the data coming back after awaiting the downstream query doesn't correspond to that expected type (see https://github.com/47ng/prisma-field-encryption/issues/80#issuecomment-1699676730).

There doesn't seem to be a way for the extension to know that the downstream query will change the returned data type, as it seems to be done by the Prisma client when resolving SQL queries, and the type system won't help us here as this is a runtime issue.

Feel free to escalate this as an issue if it seems like something that's not expected. In the meantime, I'll mark my extension as not compatible with Fluent queries.

### How to reproduce (optional)

See issue: https://github.com/47ng/prisma-field-encryption/issues/80

### Expected behavior (optional)

The extension query parameters (in my case, `$allOperations`) are somehow passed some runtime information about the returned data type when running in a Fluent context, allowing extensions to adapt.

### Information about Prisma Schema, Client Queries and Environment (optional)

See issue: https://github.com/47ng/prisma-field-encryption/issues/80

**Comment by nurul3101** [comment]
Hello @franky47 👋 

Thank you for providing a detailed description of the issue and providing a reproduction.

I would recommend opening a [GitHub Issue](https://github.com/prisma/prisma/issues/new?assignees=&labels=kind%2Fbug&template=bug_report.yml) so that our engineering team can have a look.

Meanwhile, I'll go through the [reported issue](https://github.com/47ng/prisma-field-encryption/issues/80) to check if I am able to reproduce this behaviour.

**Reply by franky47** [reply]
FYI, I have added an integration test that shows the problem in the extension repo:

1. Clone https://github.com/47ng/prisma-field-encryption
2. Install dependencies with `yarn install`
3. Build the package with `yarn build`
4. Run integration tests with `yarn test:integration`

The incriminating test (showcasing that it does not work with the Fluent API) is here:
https://github.com/47ng/prisma-field-encryption/blob/62055f48fbdc392711bc6dd76fcdaef738a0464d/src/tests/integration.test.ts#L399-L404

If you enable debugging logs with `DEBUG="prisma-field-encryption:*" yarn test:integration`, it shows that the received query expects a User.findUnique operation, but the returned data is an array of Posts.

**Comment by P4sca1** [comment]
I opened an issue for this problem:
https://github.com/prisma/prisma/issues/21988

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
